File: core/adaptive/runtime_adaptation_engine.py
from __future__ import annotations

from typing import Any, Dict, List

from core.adaptive.extraction_fallback_engine import build_extraction_fallback_chain
from core.adaptive.interaction_recovery_engine import recover_interaction_flow
from core.adaptive.modal_recovery_engine import recover_modal_runtime
from core.adaptive.pagination_recovery_engine import recover_pagination_flow


def run_runtime_adaptation(
    url: str,
    dom_nodes: List[Dict[str, Any]],
    html: str,
    interactions: List[Dict[str, Any]],
    primary_selector: str,
    page: Any = None,
) -> Dict[str, Any]:
    fallback = build_extraction_fallback_chain(
        primary_selector,
        dom_nodes,
        html,
    )
    interaction_recovery = recover_interaction_flow(
        interactions,
        dom_nodes,
        html,
    )
    modal_recovery = recover_modal_runtime(page, html)
    pagination_recovery = recover_pagination_flow(
        primary_selector,
        html,
    )

    return {
        "url": url,
        "fallback": fallback,
        "interaction_recovery": interaction_recovery,
        "modal_recovery": modal_recovery,
        "pagination_recovery": pagination_recovery,
        "bounded": True,
    }

File: src/adaptive/runtimeAdaptationEngine.ts
/**
 * Converted from Python: core/adaptive/runtime_adaptation_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import { buildExtractionFallbackChain } from "./extractionFallbackEngine.js";
import { recoverInteractionFlow } from "./interactionRecoveryEngine.js";
import { recoverModalRuntime } from "./modalRecoveryEngine.js";
import { recoverPaginationFlow } from "./paginationRecoveryEngine.js";

export function runRuntimeAdaptation(url: any, dom_nodes: any, html: any, interactions: any, primary_selector: any, page: any = null): any {
  var fallback: any = buildExtractionFallbackChain(primary_selector, dom_nodes, html);
  var interaction_recovery: any = recoverInteractionFlow(interactions, dom_nodes, html);
  var modal_recovery: any = recoverModalRuntime(page, html);
  var pagination_recovery: any = recoverPaginationFlow(primary_selector, html);
  return {"url": url, "fallback": fallback, "interaction_recovery": interaction_recovery, "modal_recovery": modal_recovery, "pagination_recovery": pagination_recovery, "bounded": true};
}
export { buildExtractionFallbackChain, recoverInteractionFlow, recoverModalRuntime, recoverPaginationFlow };
